Robus/src/robus.c

```diff
@@ -114,94 +101,118 @@ void Robus_ModulesClear(void)
-
-    for (unsigned char branch = 0; branch < NO_BRANCH; branch++)
+    // Check if ACK needed
+    if (msg->header.target_mode == IDACK)
     {
-        ctx.detection_mode = MASTER_DETECT;
-        if (Detect_PokeBranch(branch))
+        // Check if it is a localhost message
+        if (Recep_ModuleConcerned(&msg->header) && (msg->header.target != DEFAULTID))
+        {
+            Transmit_SendAck();
+            ctx.ack = 0;
+        }
+        else
         {
-            // Someone reply to our poke!
-            // loop while the line is released
-            int module_number = 0;
-            while ((ctx.detection.keepline != NO_BRANCH) & (module_number < 1024))
+            // ACK needed, change the state of state machine for wait a ACK
+            LuosHAL_SetIrqState(false);
+            ctx.data_cb = Recep_CatchAck;
+            LuosHAL_SetIrqState(true);
+            volatile int time_out = 0;
+            while (!ctx.ack & (time_out < (120 * (1000000 / ctx.baudrate))))
             {
-                if (Transmit_SetExternID(vm, IDACK, DEFAULTID, newid++))
+                time_out++;
+            }
+            status_t status;
+            status.unmap = ctx.ack;
+            if ((status.rx_error) | (status.identifier != 0xF))
+            {
+                if (ctx.ack && status.identifier != 0xF)
+                {
+                    // This is probably a part of another message
+                    // Send it to header
+                    LuosHAL_SetIrqState(false);
+                    ctx.data_cb = Recep_GetHeader;
+                    LuosHAL_SetIrqState(true);
+                    Recep_GetHeader(&ctx.ack);
+                }
+                if (nbr_nak_retry < NBR_NAK_RETRY)
+                {
+                    Recep_Timeout();
+                    for (volatile unsigned int tempo = 0; tempo < (COLLISION_TIMER * (nbr_nak_retry)); tempo++)
+                        ;
+                    goto ack_restart;
+                }
+                else
                 {
-                    // set extern id fail
-                    // remove this id and stop topology detection
-                    newid--;
-                    break;
+                    // Set the dead module ID into the VM
+                    vm->dead_module_spotted = msg->header.target;
+                    fail = 1;
                 }
-                module_number++;
-                // wait for some us
-                for (volatile unsigned int i = 0; i < (2 * TIMERVAL); i++)
-                    ;
             }
+            ctx.ack = 0;
         }
     }
-    ctx.detection_mode = NO_DETECT;
-
-    return newid - 1;
+    // localhost management
+    if (Recep_ModuleConcerned(&msg->header))
+    {
+        LuosHAL_SetIrqState(false);
+        // Secure the message memory by copying it into msg buffer
+        for (int i = 0; i < (sizeof(header_t) + msg->header.size + 2); i++)
+        {
+            ctx.data_cb(&msg->stream[i]);
+        }
```

Comment:
Instead af calling the reception state machine. allocate the message size and copy the message into the allocator.